src/Semaphore.sol

```diff
@@ -139,8 +162,8 @@ contract Semaphore is IWorldID, SemaphoreCore, Verifier, SemaphoreGroups {
         RootHistory memory rootData = rootHistory[root];
 
         if (
-            rootData.groupId != groupId ||
-            block.timestamp - rootData.timestamp > ROOT_HISTORY_EXPIRY
+                rootData.groupId != groupId ||
+                (root != latestRoots[groupId] && block.timestamp - rootData.timestamp > ROOT_HISTORY_EXPIRY)
```

Comment:
It should be possible to do this using only a single state read when it matches the latest root. We need to be careful though by default `latestRoot[x] == 0` for any `x`.